### PR TITLE
Implement lint that checks for unidiomatic unwrap() (closes #1770)

### LINKED ISSUE #1770: Lint for if x.is_ok/some { x.unwrap() }
I recently came across this in a project where I had the following code from when I was just starting to learn Rust. I found it by using pedantic mode and changing unwraps and was surprised it wasn't caught by default.

Here it is simplified:

```rust
if x.is_some() {
    let y = x.unwrap();
    // Use y
}
```
This can be better written as
```rust
if let Some(y) = x {
    // Use y
}
```

### PR DESCRIPTION
**Pull request by fanzier**
In the process I moved some logic to determine which variables are
mutated in an expression to utils/usage.rs.

cc @oli-obk since we talked about this yesterday

### CODE REVIEW

**Comment by fanzier** [comment]
The test fails with *node_id_to_type: no type for node `expr value (id=1627)`*.
The type is requested in `unwrap.rs:75` but how can the expression not have a type?

**Comment by mati865** [comment]
This PR needs rebase.
Pinging @oli-obk for review

File: clippy_lints/src/unwrap.rs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+/// Could be written:
+///
+/// ```rust
+/// if let Some(value) = option {
+///     do_something_with(value)
+/// }
+/// ```
+declare_clippy_lint! {
+    pub UNNECESSARY_UNWRAP,
+    complexity,
```
Comment (oli-obk) [inline clippy_lints/src/unwrap.rs]: let's put it in the `nursery` for now until we have some data on larger projects

File: clippy_lints/src/unwrap.rs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            // find `unwrap[_err]()` calls:
+            if_chain! {
+                if let Expr_::ExprMethodCall(ref method_name, _, ref args) = expr.node;
+                if let Expr_::ExprPath(QPath::Resolved(None, ref path)) = args[0].node;
+                if ["unwrap", "unwrap_err"].contains(&&*method_name.name.as_str());
+                let call_to_unwrap = method_name.name == "unwrap";
+                if let Some(unwrappable) = self.unwrappables.iter()
+                    .find(|u| u.ident.def == path.def && call_to_unwrap == u.safe_to_unwrap);
+                then {
+                    self.cx.span_lint_note(
```
Comment (oli-obk) [inline clippy_lints/src/unwrap.rs]: clippy has various custom methods for linting, please use them. Most likely you'll want `span_lint_and_then`

Also don't use a `note`, but a `label`, as that will probably produce a prettier error message.

Comment (fanzier) [reply clippy_lints/src/unwrap.rs]: Cool, `span_label` is much prettier :)

**Comment by fanzier** [comment]
Rebased and addressed your comments

**Comment by oli-obk** [comment]
Thanks!

**Pull request merged**